src/geometry/manifold/algebra/smooth_functions.lean

```diff
@@ -0,0 +1,233 @@
+  [has_continuous_mul G] [charted_space H' G]
+  [smooth_manifold_with_corners I' G] [has_smooth_mul I' G] :
+  comm_monoid C∞(I, N; I', G) :=
+{ one_mul := λ a, by ext; exact one_mul _,
+  mul_one := λ a, by ext; exact mul_one _,
+  mul_comm := λ a b, by ext; exact mul_comm _ _,
+  ..smooth_map_semigroup,
+  ..smooth_map.has_one }
+
+@[to_additive]
+instance smooth_map_group {G : Type*} [group G] [topological_space G] [topological_group G]
+  [charted_space H' G] [smooth_manifold_with_corners I' G] [lie_group I' G] :
+  group C∞(I, N; I', G) :=
+{ inv := λ f, ⟨λ x, (f x)⁻¹, smooth_inv.comp f.smooth⟩,
+  mul_left_inv := λ a, by ext; exact mul_left_inv _,
+  ..smooth_map_monoid }
+
+@[to_additive]
+instance smooth_map_comm_group {G : Type*} [comm_group G] [topological_space G]
+  [topological_group G] [charted_space H' G] [smooth_manifold_with_corners I' G] [lie_group I' G] :
+  comm_group C∞(I, N; I', G) :=
+{ ..smooth_map_group,
+  ..smooth_map_comm_monoid }
+
+end group_structure
+
+section ring_structure
+
+/-!
+### Ring stucture
+
+In this section we show that continuous functions valued in a topological ring `R` inherit
+a structure of ring.
+-/
+
+instance smooth_map_semiring {R : Type*} [semiring R] [topological_space R] [topological_semiring R]
+  [charted_space H' R] [smooth_manifold_with_corners I' R] [smooth_semiring I' R] :
+  semiring C∞(I, N; I', R) :=
+{ left_distrib := λ a b c, by ext; exact left_distrib _ _ _,
+  right_distrib := λ a b c, by ext; exact right_distrib _ _ _,
+  zero_mul := λ a, by ext; exact zero_mul _,
+  mul_zero := λ a, by ext; exact mul_zero _,
+  ..smooth_map_add_comm_monoid,
+  ..smooth_map_monoid }
+
+instance smooth_map_ring {R : Type*} [ring R] [topological_space R] [topological_ring R]
+  [charted_space H' R] [smooth_manifold_with_corners I' R] [smooth_ring I' R] :
+  ring C∞(I, N; I', R) :=
+{ ..smooth_map_semiring,
+  ..smooth_map_add_comm_group, }
+
+instance smooth_map_comm_ring {R : Type*} [comm_ring R] [topological_space R] [topological_ring R]
+  [charted_space H' R] [smooth_manifold_with_corners I' R] [smooth_ring I' R] :
+  comm_ring C∞(I, N; I', R) :=
+{ ..smooth_map_semiring,
+  ..smooth_map_add_comm_group,
+  ..smooth_map_comm_monoid,}
+
+end ring_structure
+
+section semimodule_structure
+
+/-!
+### Semiodule stucture
+
+In this section we show that continuous functions valued in a topological semimodule `M` over a
+topological semiring `R` inherit a structure of semimodule.
+-/
+
+instance smooth_map_has_scalar
+  {R : Type*} [semiring R] [topological_space R]
+  [charted_space H' R] [smooth_manifold_with_corners I' R]
+  {M : Type*} [topological_space M] [add_comm_monoid M]
+  [semimodule R M] [topological_semimodule R M]
+  [charted_space H'' M] [smooth_manifold_with_corners I'' M] [smooth_semimodule I' I'' R M] :
+  has_scalar R C∞(I, N; I'', M) :=
+⟨λ r f, ⟨r • f, (@smooth_const _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ I' _ _ _ _).smul f.smooth⟩⟩
+/- I don't know how to fix this -/
```

Comment:
Dangerous instance I don't know how to solve. I will do some research to find out but help is most welcome

Reply:
The instance conclusion does not mention `H'` at all, so Lean can never guess what `H'` you want to use. 

What are you trying to say here? I mean, is your goal to use it for `R` the base field (in which case there will be no problem since the field can be found by looking at `I` or `I''`), or do you have a more fancy application in mind?

Reply:
Yeah sure I see why it is dangerous, but all the same I did not know how to find a solution. No not necessarily the base field (I mean often one considers real functions on a complex manifold) but a field so I will fix it similarly if there is no other solution.

Reply:
Then you can just use `H' = R` and `model_with_corners_self 𝕜 R` instead of `I'`, and everything should be fine.